rootfs: support EROFS filesystem

yaoyinnan opened the pull request:
For kata containers, rootfs is used in the read-only way.
EROFS can noticably decrease metadata overhead.

Undertake https://github.com/kata-containers/kata-containers/pull/2586 work from @hsiangkao.

On the basis of supporting the EROFS file system, it supports using the config parameter to switch the file system used by rootfs.

Fixes: #6063

Signed-off-by: Gao Xiang <hsiangkao@linux.alibaba.com>
Signed-off-by: yaoyinnan <yaoyinnan@foxmail.com>

hsiangkao commented on src/runtime/Makefile:
```diff
@@ -112,6 +112,7 @@ KERNELDIR := $(PKGDATADIR)
 
 IMAGEPATH := $(PKGDATADIR)/$(IMAGENAME)
 INITRDPATH := $(PKGDATADIR)/$(INITRDNAME)
+DEFROOTFSTYPE := \"ext4\"
```
I'd suggest to switch to erofs by default if possible. I think the benefit is enough

hsiangkao replied:
we can show benefits explicitly here, but it can be done in a separated PR

yaoyinnan replied:
I switch EROFS as default rootfs filesystem.

I think might be better to show the benefits of EROFS in image_builder.sh. I will add it when the documentation is complete later.

hsiangkao commented on tools/osbuilder/image-builder/image_builder.sh:
```diff
@@ -476,6 +476,80 @@ and the rootfs is built with SELINUX=yes"
+	touch "${mount_dir}/etc/machine-id"
+
+	readonly fsimage="$(mktemp)"
+	mkfs.erofs "${fsimage}" "${mount_dir}"
```
we need to add "-Enoinline_data" to enable FSDAX for all files for now (until v6.1).

yaoyinnan replied:
added

hsiangkao commented on tools/osbuilder/image-builder/image_builder.sh:
```diff
@@ -476,6 +476,80 @@ and the rootfs is built with SELINUX=yes"
+	local image="$2"
+	local fs_type="$3"
+	local block_size="$4"
+  local agent_bin="$5"
```
inconsistant tab indent

yaoyinnan replied:
fixed

hsiangkao commented on tools/osbuilder/image-builder/image_builder.sh:
```diff
@@ -566,14 +640,24 @@ main() {
+
+		# mkfs.erofs accepts an src root dir directory as an input
+		# rather than some device, so no need to guess the device dest size first.
+    rootfs_img_size="$?"
```
here as well

yaoyinnan replied:
fixed

studychao commented:
Thanks! @yaoyinnan , could you also add support in runtime-rs too?

yaoyinnan commented:
> Thanks! @yaoyinnan , could you also add support in runtime-rs too?

I added support for EROFS in runtime-rs.

bergwolf commented on src/runtime/Makefile:
```diff
@@ -112,6 +112,12 @@ KERNELDIR := $(PKGDATADIR)
+ROOTFSTYPE_EXT4 := \"ext4\"
+ROOTFSTYPE_XFS := \"xfs\"
+ROOTFSTYPE_EROFS := \"erofs\"
+DEFROOTFSTYPE := $(ROOTFSTYPE_EROFS)
```
Why is the default rootfs type changed?

hsiangkao replied:
Let's use `ext4` for now until enough reasons to switch to `erofs`.
We need to show the benefits first.

yaoyinnan replied:
I temporarily switched back to `ext4` as the default.
We need to demonstrate the benefits first before considering switching `erofs` as the default.

liubin commented on src/runtime/virtcontainers/hypervisor.go:
```diff
@@ -112,6 +126,66 @@ var commonVirtioblkKernelRootParams = []Param{ //nolint: unused, deadcode, varch
 	{"rootfstype", "ext4"},
 }
 
+func NvdimmKernelRootParams(rootfstype string) []Param {
```
Please add tests for these new NvdimmKernelRootParams/NvdimmNoDAXKernelRootParams/VirtioblkKernelRootParams  
 functions.

yaoyinnan replied:
I added unit tests to cover all rootfs type corresponding kernel root parameters.

liubin commented on tools/osbuilder/image-builder/image_builder.sh:
```diff
@@ -476,6 +476,80 @@ and the rootfs is built with SELINUX=yes"
+create_erofs_rootfs_image() {
+	local rootfs="$1"
+	local image="$2"
+	local fs_type="$3"
```
`fs_type` is not used?

yaoyinnan replied:
deleted

liubin commented on tools/osbuilder/image-builder/image_builder.sh:
```diff
@@ -476,6 +476,80 @@ and the rootfs is built with SELINUX=yes"
+and the rootfs is built with SELINUX=yes"
+		fi
+	fi
+
```
Add 
```bash
	sync
	OK "rootfs copied"
``` 
like what `create_rootfs_image` has done?

yaoyinnan replied:
added

liubin commented on tools/osbuilder/image-builder/image_builder.sh:
```diff
@@ -476,6 +476,80 @@ and the rootfs is built with SELINUX=yes"
+
+	local img_size="$(stat -c"%s" ${fsimage})"
+	local img_size_mb="$((((${img_size} + 1048576) / 1048576) + 1 + ${rootfs_start}))"
+	create_disk "${image}" "${img_size_mb}" "ext4" "${rootfs_start}"
```
Should here `ext4` be `$fs_type`?

yaoyinnan replied:
EROFS is only used for formatting and cannot be used for partitioning. Ext4 is required for partitioning.

hsiangkao replied:
IMHO, it seems that we could support `GNU parted` later, it seems that `f2fs` was just supported in 2020:
https://git.savannah.gnu.org/gitweb/?p=parted.git;a=commit;h=c6b61814cd4cf958e12d35a36184ff7d767e57d9

yaoyinnan commented on tools/osbuilder/image-builder/image_builder.sh:
```diff
@@ -476,6 +476,80 @@ and the rootfs is built with SELINUX=yes"
 	rmdir "${mount_dir}"
 }
 
+create_erofs_rootfs_image() {
```
I stripped down the repetitive code.
However, since the construction process of EROFS is different from that of ext4 and xfs, too much redundant judgment logic will be added in the same method, and `create_erofs_rootfs_image()` is still used alone.

yaoyinnan replied:
I encapsulated the operations of `systemd` and `SELinux` into `setup_systemd()` and `setup_selinux()` methods.

yaoyinnan commented on src/runtime-rs/crates/hypervisor/src/kernel_param.rs:
```diff
@@ -67,39 +68,113 @@ impl KernelParams {
-                        value: "ext4".to_string(),
-                    },
-                ]
+                match rootfs_type {
```
I stripped down the repetitive code.

bergwolf reviewed (approved):
lgtm, thanks!

hsiangkao commented on tools/osbuilder/image-builder/image_builder.sh:
```diff
@@ -473,7 +491,50 @@ and the rootfs is built with SELINUX=yes"
+	info "Setup systemd"
+	setup_systemd "${mount_dir}"
+
+  readonly fsimage="$(mktemp)"
```
inconsistent indentation.

yaoyinnan replied:
fixed.

yaoyinnan commented:
> @yaoyinnan there are two errors:
> 
> one is
> 
> > ERROR: Document docs/how-to/how-to-use-erofs-build-rootfs.md is not referenced
> 
> and since https://github.com/kata-containers/kata-containers/pull/6163/files#diff-b53e92e39d4d8de2382e1c11820bd1156810d063c6b2a0b1b537b99af3c8320bR1 has been merged, you need rebase main branch to increase the kernel config version.

I added references, and switched kernel config version.

yaoyinnan commented:
@bergwolf @liubin @hsiangkao I modified some errors with reference to the comments, please review it for me.

hsiangkao commented on docs/how-to/how-to-use-erofs-build-rootfs.md:
```diff
@@ -0,0 +1,100 @@
+## Performance comparison
+|                 | EROFS | EXT4 | XFS |
+|-----------------| --- | --- | --- |
+| Image Size [MB] | 106 | 256 | 126 |
```
Compressed Image size (obvious compressed EROFS is smaller than 106MiB.)

yaoyinnan replied:
I noted that this is the uncompressed image size.

hsiangkao commented on docs/how-to/how-to-use-erofs-build-rootfs.md:
```diff
@@ -0,0 +1,100 @@
+#### Source install
+[https://git.kernel.org/pub/scm/linux/kernel/git/xiang/erofs-utils.git](https://git.kernel.org/pub/scm/linux/kernel/git/xiang/erofs-utils.git)
+
+##### Compile dependencies
```
Compilation dependencies

liubin commented on docs/how-to/how-to-use-erofs-build-rootfs.md:
```diff
@@ -0,0 +1,100 @@
+$ pushd kata-containers/tools/osbuilder/image-builder
+$ commit="$(git log --format=%h -1 HEAD)"
+$ date="$(date +%Y-%m-%d-%T.%N%z)"
+$ image="kata-containers-${date}-${commit}"
```
Nit: Adding `erofs` in the file name will help to understand.

yaoyinnan replied:
added.

liubin commented on docs/how-to/how-to-use-erofs-build-rootfs.md:
````diff
@@ -0,0 +1,100 @@
+$ popd
+```
+
+### Build and install guest kernel images
````
Is this required? If not, consider removing it or adding `OPTIONAL`.

Or could you describe why this step is required?

IMO, after this PR has been merged, users will not need to build the kernel.

yaoyinnan replied:
Indeed, build guest kerel is not needed here. I removed this part.

c3d commented on docs/how-to/how-to-use-erofs-build-rootfs.md:
````diff
@@ -0,0 +1,100 @@
+```shell
+$ apt install erofs-utils
+```
+
````
Please add instructions for Fedora (same package name):

```
$ dnf install erofs-utils
```

(Checked on Fedora 37)

yaoyinnan replied:
added.

liubin commented on docs/how-to/how-to-use-erofs-build-rootfs.md:
````diff
@@ -0,0 +1,100 @@
+$ popd
+```
+
+### Build the runtime
````
Ditto.

yaoyinnan replied:
I adjusted to modify the rootfs_type of /etc/kata-containers/configuration.toml using sed command.

c3d commented on docs/how-to/how-to-use-erofs-build-rootfs.md:
```diff
@@ -0,0 +1,100 @@
+$ pushd kata-containers/src/runtime
+$ make && sudo -E "PATH=$PATH" make install
+$ sudo mkdir -p /etc/kata-containers/
+$ sudo install -o root -g root -m 0640 /usr/share/defaults/kata-containers/configuration.toml /etc/kata-containers
```
I don't think we should recommend the step of installing stuff under `/etc` unless we actually change the file. The philosophy is that `defaults` contains the default configuration, and that you can override it with local changes that you put under `/etc/kata-containers`. On the other hand, if you have not made changes, but copied an "old" `configuration.toml` under `/etc/kata-containers`, then perform an upgrade that introduces a change in the default `configuration.toml`, then you won't get the benefit of that change unless you copy again.

yaoyinnan replied:
I adjusted to modify the `rootfs_type` of `/etc/kata-containers/configuration.toml` using `sed` command.

c3d commented on src/runtime-rs/crates/hypervisor/src/kernel_param.rs:
```diff
@@ -67,43 +69,37 @@ impl KernelParams {
-                        value: "ext4".to_string(),
-                    },
-                ]
+    pub(crate) fn new_rootfs_kernel_params(rootfs_driver: &str, rootfs_type: &str) -> Result<Self> {
```
Thinking long-term here, have you considered how to achieve that result when you cannot easily change the command-line parameters (e.g. in the confidential containers case)?

hsiangkao replied:
Personally I'd suggest it as an unblocked stuff here but yeah confidential containers cases are quite different and  it needs to be resolved in the guest itself. It'd be better to be addressed in another PR.

Currently we'd like to use EROFS as vm rootfs for our production. And we think our attempt on this is useful to the whole kata communities for referenece ;-)

c3d replied:
Yes, it's a question, it is definitely not part of this PR.

hsiangkao commented on docs/how-to/how-to-use-erofs-build-rootfs.md:
````diff
@@ -0,0 +1,90 @@
+On newer Debian/Ubuntu systems, it can be installed directly using the apt command.
+
+```shell
+# Debian
````
Fedora?

yaoyinnan replied:
sorry, typo

hsiangkao commented on docs/how-to/how-to-use-erofs-build-rootfs.md:
```diff
@@ -0,0 +1,90 @@
+## Guidance
+### Install the `erofs-utils`
+#### apt install
+On newer Debian/Ubuntu systems, it can be installed directly using the apt command.
```
this part is still broken if considering non-Debian ecosystem.

yaoyinnan replied:
updated

hsiangkao commented on docs/how-to/how-to-use-erofs-build-rootfs.md:
```diff
@@ -0,0 +1,90 @@
+[https://git.kernel.org/pub/scm/linux/kernel/git/xiang/erofs-utils.git](https://git.kernel.org/pub/scm/linux/kernel/git/xiang/erofs-utils.git)
+
+##### Compile dependencies
+If you need to support the `Lz4` compression feature, you need `Lz4 1.8.0+`, and `Lz4 1.9.3+` is strongly recommended.
```
If you need to enable the `Lz4` compression feature, `Lz4 1.8.0+` is required, and `Lz4 1.9.3+` is strongly recommended.

yaoyinnan replied:
I changed to your description.

hsiangkao commented on docs/how-to/how-to-use-erofs-build-rootfs.md:
```diff
@@ -0,0 +1,90 @@
+
+## Guidance
+### Install the `erofs-utils`
+#### apt install
```
why we still have `apt install` here?

yaoyinnan replied:
Sorry, I forgot to edit earlier.

yaoyinnan commented:
/test

yaoyinnan commented:
/test-fc

yaoyinnan commented:
/test-fc

yaoyinnan commented:
/test-ubuntu

yaoyinnan commented:
/test

yaoyinnan commented:
/test-arm

yaoyinnan commented:
/test

Apokleos commented on src/runtime-rs/crates/hypervisor/src/kernel_param.rs:
```diff
@@ -7,6 +7,8 @@
 use anyhow::{anyhow, Result};
 
 use crate::{VM_ROOTFS_DRIVER_BLK, VM_ROOTFS_DRIVER_PMEM};
+use crate::{VM_ROOTFS_FILESYSTEM_EROFS, VM_ROOTFS_FILESYSTEM_EXT4, VM_ROOTFS_FILESYSTEM_XFS};
+use crate::{VM_ROOTFS_ROOT_BLK, VM_ROOTFS_ROOT_PMEM};
```
The two `use`s merged into one line `use crate::{...}` will be better.

yaoyinnan replied:
@Apokleos Thanks for your review! I merged use crate.

Apokleos commented on src/runtime-rs/crates/hypervisor/src/kernel_param.rs:
```diff
@@ -67,43 +69,37 @@ impl KernelParams {
+    pub(crate) fn new_rootfs_kernel_params(rootfs_driver: &str, rootfs_type: &str) -> Result<Self> {
+        let mut params = vec![];
+
+        if rootfs_driver == VM_ROOTFS_DRIVER_PMEM {
+            params.push(Param::new("root", VM_ROOTFS_ROOT_PMEM));
+            if rootfs_type == VM_ROOTFS_FILESYSTEM_EXT4 || rootfs_type == VM_ROOTFS_FILESYSTEM_XFS {
+                params.push(Param::new(
+                    "rootflags",
+                    "dax,data=ordered,errors=remount-ro ro",
+                ));
+            } else if rootfs_type == VM_ROOTFS_FILESYSTEM_EROFS {
+                params.push(Param::new("rootflags", "dax ro"));
+            } else {
+                return Err(anyhow!("Unsupported rootfs type"));
```
Reduce the use of `if {} else if {}` and a more elegant way is `match` {...}

yaoyinnan replied:
@Apokleos Thanks for your review! I switched `if {} else if {}` to `match{...}`.

yaoyinnan commented:
/test

hsiangkao commented:
``
--- FAIL: TestQemuArm64AppendImage (0.00s)
    qemu_arm64_test.go:113: 
        	Error Trace:	/home/jenkins/workspace/kata-containers-2.0-ubuntu20.04-ARM-unit/go/src/github.com/kata-containers/kata-containers/src/runtime/virtcontainers/qemu_arm64_test.go:113
        	Error:      	Received unexpected error:
        	            	unsupported rootfs type
        	Test:       	TestQemuArm64AppendImage
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x120 pc=0xe568bc]

goroutine 788 [running]:
testing.tRunner.func1.2({0xf87e00, 0x1b7e0b0})
	/usr/local/go/src/testing/testing.go:1396 +0x278
testing.tRunner.func1()
	/usr/local/go/src/testing/testing.go:1399 +0x42c
panic({0xf87e00, 0x1b7e0b0})
	/usr/local/go/src/runtime/panic.go:890 +0x260
github.com/kata-containers/kata-containers/src/runtime/virtcontainers.TestQemuArm64AppendImage(0xc0007c9040)
	/home/jenkins/workspace/kata-containers-2.0-ubuntu20.04-ARM-unit/go/src/github.com/kata-containers/kata-containers/src/runtime/virtcontainers/qemu_arm64_test.go:114 +0x1bc
testing.tRunner(0xc0007c9040, 0x1160f18)
	/usr/local/go/src/testing/testing.go:1446 +0x18c
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:1493 +0x568
``
we need to skip some tests due to qemu-img limitation? @yaoyinnan @liubin

yaoyinnan commented:
/test

yaoyinnan commented:
/test

yaoyinnan commented:
/test-s390x

yaoyinnan commented:
/test-s390x

yaoyinnan commented:
/test-s390x

yaoyinnan commented:
/test-s390x-main

hsiangkao commented on src/runtime/virtcontainers/qemu_s390x.go:
```diff
@@ -83,7 +83,11 @@ func newQemuArch(config HypervisorConfig) (qemuArch, error) {
 
 	if config.ImagePath != "" {
-		q.kernelParams = append(q.kernelParams, commonVirtioblkKernelRootParams...)
+		kernelParams, err := GetKernelRootParams(config.RootfsType, false, false)
```
here should be 
`kernelParams, err := GetKernelRootParams(config.RootfsType, true, false)` ?

yaoyinnan replied:
It was this problem that caused the failure of the s390x job. I have modified it correctly.

yaoyinnan commented:
/test

yaoyinnan commented:
/test

yaoyinnan commented:
/test
